src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/StringSearchValuesHelper.cs

```diff
@@ -62,47 +62,117 @@ public static bool StartsWith<TCaseSensitivity>(ref char matchStart, int lengthR
-            public static bool AtLeast8CharsOrUnknown => true;
+            public readonly string Value;
+            public readonly nint SecondReadByteOffset;
+            public readonly Vector256<ushort> Value256;
+            public readonly Vector256<ushort> ToUpperMask256;
+
+            public readonly ulong Value64_0 => Value256.AsUInt64()[0];
+            public readonly ulong Value64_1 => Value256.AsUInt64()[1];
+            public readonly uint Value32_0 => Value256.AsUInt32()[0];
+            public readonly uint Value32_1 => Value256.AsUInt32()[1];
+
+            public readonly ulong ToUpperMask64_0 => ToUpperMask256.AsUInt64()[0];
+            public readonly ulong ToUpperMask64_1 => ToUpperMask256.AsUInt64()[1];
+            public readonly uint ToUpperMask32_0 => ToUpperMask256.AsUInt32()[0];
+            public readonly uint ToUpperMask32_1 => ToUpperMask256.AsUInt32()[1];
+
+            public SingleValueState(string value, bool ignoreCase)
+            {
+                Debug.Assert(value.Length >= 2);
+
+                Value = value;
+
+                // We precompute vectors specifc to this value to speed up later comparisons.
+                // We group values depending on their length (2-3, 4-8, 9-16).
+                // For any of those lengths, we can load the whole value with two overlapped reads (e.g. 2x 8 characters for lengths 9-16).
+                // For a string "Hello World", we would load
+                //              [Hello Wo]
+                //                 [lo World]
+                // SecondReadByteOffset: 6 bytes (3 characters)
+                // We then precompute a mask that converts any potential input to the uppercase variant, specific to this value.
+                // We must ensure that the ASCII letter mask only applies to the letters, not the space character.
+                // Value256:       [HELLO WOLO WORLD] (note that the value is already converted to uppercase if we're ignoring casing)
+                // ToUpperMask256: [xxxxx xxxx xxxxx] (x = ~0x20 for ASCII letters, 0xFFFF otherwise)
+                //
+                // Given a potential match, we can now confirm whether we found a match by loading the candidate in the same way and appliying this mask:
+                // Vector256 input = [Vector128.Load(candidate), Vector128.Load(candidate + 6 bytes)];
+                // bool matches = (input & ToUpperMask256) == Value256;
+
+                if (value.Length <= 16)
+                {
+                    if (value.Length > 8)
+                    {
+                        SecondReadByteOffset = (value.Length - 8) * sizeof(char);
+                        Value256 = Vector256.Create(
+                            Vector128.LoadUnsafe(ref value.GetRawStringDataAsUInt16()),
+                            Vector128.LoadUnsafe(ref Unsafe.AddByteOffset(ref value.GetRawStringDataAsUInt16(), SecondReadByteOffset)));
+                    }
```

Comment:
Would it make sense to move this block to within the >= 4 block? It'd mean an extra comparison in the > 8 case but would save a comparison in the < 4 case.